nelox/scanner.py

```diff
@@ -0,0 +1,117 @@
+from typing import List
+
+from .nelox_token import Token
+from .token_type import TokenType
+
+
+def is_alpha_numeric(c):
+    return is_alpha(c) or is_digit(c)
+
+
+def is_alpha(c):
+    return c.isalpha() or c == '_'
+
+
+def is_digit(c):
+    return '0' <= c <= '9'
+
+
+class Scanner:
+
+    def __init__(self, source: str):
+        self.source: str = source
+        self.tokens: List[Token] = []
+        self.start: int = 0
+        self.current: int = 0
+        self.line: int = 1
+
+    def scan_tokens(self) -> List[Token]:
+        while not self.is_at_end():
+            self.start = self.current
```

Comment:
Hang on, this looks suspicious.  How does this work if the user is using `next_token` themselves?